app/controllers/delivery_areas_controller.rb

```diff
@@ -1,79 +1,61 @@
 class DeliveryAreasController < ApplicationController
-  before_action :set_delivery_area, only: [:show, :edit, :update, :destroy]
-   before_action :set_city, only: [:new, :create]
-  def index
-    @delivery_areas = DeliveryArea.all
+  before_filter :authenticate_user!
+  load_and_authorize_resource
+   before_action :set_delivery_area, only: [:edit, :update, :destroy]
+   before_action :set_city
+   before_action :set_hotel, only:[:edit, :new, :create, :update, :destroy]
 
-  end
-
-  def show
-  end
-
-  def new   
-    @hotel = Hotel.find(params[:hotel_id])
+  def new
     @delivery_area = DeliveryArea.new
-    @city=City.all
+    @delivery_areas = DeliveryArea.where(hotel_id: params[:hotel_id])
   end
 
-  # GET /delivery_areas/1/edit
-  def edit
-  end
   def create
-    area_arr=params[:area_id]
-    if area_arr.nil?
-    else
-      area_arr.each do |id|
+    area_arr=params[:area_id] #get multiple areas
+    area_arr.each do |id|
       @delivery_area = DeliveryArea.new(delivery_area_params)
+      @delivery_area.city = City.find(params[:city_id])
       @delivery_area.area_id = id
       @delivery_area.hotel_id = params[:hotel_id]
       area = Area.find(id)
       area.hotels << Hotel.find(params[:hotel_id])
-      area.save!      
-    end  
-
-    respond_to do |format|
+      area.save!
```

Comment:
check out background processing and observers, area saving should ideally be a background job.